Fixed  #24495 -- Allow unsaved instance in ForeignKey check to be bypassed

kaedroho opened the pull request:
This pull request implements an ``allow_unsaved_instance_assignment`` flag which can be used by custom related fields to opt-out of the new check implemented in: https://github.com/django/django/commit/5643a3b51be338196d0b292d5626ad43648448d3

collinanderson commented:
buildbot, test this please.

timgraham commented on tests/one_to_one/models.py:
```diff
@@ -118,3 +118,18 @@ class Director(models.Model):
+
+class UnsavedOneToOneField(models.OneToOneField):
+    # A OneToOneField which can point to an unsaved object
+    unsaved_instances_allowed = True
```
Could you add documentation for this? I'd suggest the name `allow_unsaved_instance_assignment`

kaedroho replied:
Thanks @timgraham. I've changed the name and added a note to releases/1.8. Shall I add further documentation to https://docs.djangoproject.com/en/1.7/ref/models/fields/ ?

timgraham replied:
Yes

timgraham commented on docs/releases/1.8.txt:
```diff
@@ -713,6 +713,18 @@ Now, an error will be raised to prevent data loss::
+
+    class Book(models.Model):
+        author = UnsavedOnetoOneField(Author)
+
```
just once space between sections

timgraham commented on tests/many_to_one/models.py:
```diff
@@ -98,3 +98,18 @@ class School(models.Model):
     school = models.ForeignKey(School)
+
+
+class UnsavedForeignKey(models.ForeignKey):
```
Since it doesn't seem that the database tables are required, could we inline these classes in the test?

kaedroho replied:
Do you know of an example of this happening in another test which I can use as a reference?

timgraham replied:
https://github.com/django/django/blob/71c638fc55d5b0637e0c50a71694f3e422caa0c1/tests/model_fields/tests.py#L186-L187

timgraham commented:
Please squash your commits when you are finished with those edits, thanks.

kaedroho commented:
@timgraham I've made those changes and squashed the commits. Let me know if theres anything else.

timgraham commented:
buildbot, add to whitelist.

timgraham commented:
Should the same flag be added to [GenericForeignKey](https://github.com/django/django/blob/64b9393ae279fde1ac535348458b3ff960177225/django/contrib/contenttypes/fields.py#L253-L257)? Do you think there might be a demand to allow specifying `allow_unsaved_instance_assignment` as a field constructor argument instead of requiring a subclass?

timgraham commented:
Commit message should include the ticket and be in past tense:
```
Fixed #24495 -- Allowed unsaved model instance assignment check to be bypassed.
```
Here are some doc edits: http://dpaste.com/0WYH7W1

kaedroho commented:
Thanks, I'll make those changes this evening (GMT)

kaedroho commented:
Yep, I don't think it would hurt to add it to ``GenericForeignKey`` as well.

I've added a commit for this: https://github.com/kaedroho/django/commit/a25d95e9dfa0051677120465e74271d329b51aca

Let me know if it looks ok to you and I'll squash it in.

timgraham commented on docs/ref/contrib/contenttypes.txt:
```diff
@@ -299,6 +299,33 @@ model:
 
+    .. attribute:: GenericForeignKey.allow_unsaved_instance_assignment
+
+        Django prevents unsaved model instances from being assigned to a
```
I'd rather simply link to `ForeignKey.allow_unsaved_instance_assignment` as `OneToOne` field does. I don't expect it to be commonly used so I think it'd be better not to make a big deal about it.

timgraham commented on docs/ref/models/fields.txt:
```diff
@@ -1483,6 +1507,10 @@ that control how the relationship functions.
 
     If in doubt, leave it to its default of ``True``.
 
+.. attribute:: ForeignKey.allow_unsaved_instance_assignment
```
`OneToOneField.` (my typo)

kaedroho replied:
This looks ok to me...

timgraham replied:
Oops, ignore me. I'll delete that comment.

timgraham commented on tests/contenttypes_tests/tests.py:
```diff
@@ -257,6 +257,36 @@ class Model(models.Model):
+        """
+        #24495 - Assigning an unsaved object to GenericForeignKey
+        should be allowed when the allow_unsaved_instance_assignment
+        attribute has been set to True
```
add period

timgraham commented on tests/contenttypes_tests/tests.py:
```diff
@@ -257,6 +257,36 @@ class Model(models.Model):
+        beatles = Band(name='The Beatles')
+        john = BandMember(first_name='John', last_name='Lennon')
+
+        # This should not raise an exception as the ForeignKey
```
add period to the sentences

timgraham commented on tests/contenttypes_tests/tests.py:
```diff
@@ -257,6 +257,36 @@ class Model(models.Model):
+            first_name = models.CharField(max_length=50)
+            last_name = models.CharField(max_length=50)
+
+        # Create band and member objects
```
not necessary

timgraham commented on tests/contenttypes_tests/tests.py:
```diff
@@ -257,6 +257,36 @@ class Model(models.Model):
 
+    def test_unsaved_instance_on_generic_foreign_key_allowed_when_wanted(self):
+        """
+        #24495 - Assigning an unsaved object to GenericForeignKey
```
A bit odd that this test has a doc string and the others don't.

timgraham commented on django/contrib/contenttypes/fields.py:
```diff
@@ -250,7 +250,7 @@ def __set__(self, instance, value):
             ct = self.get_content_type(obj=value)
             fk = value._get_pk_val()
-            if fk is None:
+            if not getattr(self, 'allow_unsaved_instance_assignment', False) and fk is None:
```
is there a reason to use `getattr` instead of defining the attribute as `False` on the class?

timgraham commented:
Looks fine to squash with those updates.

timgraham commented on docs/ref/contrib/contenttypes.txt:
```diff
@@ -299,6 +299,11 @@ model:
+    .. attribute:: GenericForeignKey.allow_unsaved_instance_assignment
+
+       Works analogously to :attr:`ForeignKey.allow_unsaved_instance_assignment`.
+
```
one space between sections

kaedroho replied:
Oops, fixed.

timgraham commented on docs/ref/contrib/contenttypes.txt:
```diff
@@ -299,6 +299,11 @@ model:
 
+    .. attribute:: GenericForeignKey.allow_unsaved_instance_assignment
+
+       Works analogously to :attr:`ForeignKey.allow_unsaved_instance_assignment`.
```
need `ForeignKey.allow_unsaved_instance_assignment <django.db.models.ForeignKey.allow_unsaved_instance_assignment>` here since we aren't in the django.db.models module of the docs.

kaedroho replied:
Fixed, thanks!

timgraham commented:
Please check the test failures. Maybe the new attribute should be moved to the `Field` base class?

timgraham commented:
The issue seems to be with reverse relations. I'm unsure if the current API is sufficient. For example if you have a relation to an app you don't control and want to disable the check.

kaedroho commented:
@timgraham There was a mistake around working out which attribute holds the field with the ``allow_unsaved_instance_assignment`` attribute. This is fixed now

I've also added it to ``ForeignObject`` (and added a test for #10811 as it didn't have one).

All changes squashed

kaedroho commented on django/db/models/fields/related.py:
```diff
@@ -1856,6 +1857,7 @@ class ForeignKey(ForeignObject):
 
     rel_class = ManyToOneRel
 
+    allow_unsaved_instance_assignment = False
```
I'll remove this as ``ForeignKey`` is a subclass of ``ForeignObject``

timgraham commented:
merged in 81e1a35c364e5353d2bf99368ad30a4184fbb653, thanks!
